src/datum_utils.c

```diff
@@ -703,3 +703,16 @@ uint64_t datum_siphash_mod8(const void *src, uint64_t sz, const unsigned char ke
 	SIPHASH_DOUBLE_ROUND(v0,v1,v2,v3);
 	return (v0 ^ v1) ^ (v2 ^ v3);
 }
+
+// Uses a fixed-size buffer; positive only; digits only
+// Returns -1 on failure
+int datum_atoi_strict(const char * const s, const size_t size) {
+	if (!size) return -1;
+	int ret = 0;
+	for (size_t i = 0; i < size; ++i) {
+		if (s[i] < '0' || s[i] > '9') return -1;
+		if (ret > INT_MAX / 10) return -1;
```

Comment:
I fuzzed this function and this check fails around the edges. It doesn't work for 2 valid inputs because it only considers whether the current value of ret would overflow when multiplied by 10. However, it does not account for the addition of the last digit, which could push the result over INT_MAX.
For input '2147483648', instead of -1 we get -2147483648
For input '2147483649', instead of -1 we get -2147483647

You could set the digit before the check and subtract it. Like this:
```
	for (size_t i = 0; i < size; ++i) {
		if (s[i] < '0' || s[i] > '9') return -1;
		unsigned char digit = s[i] - '0';
		if (ret > (INT_MAX - digit) / 10) return -1;
		ret = (ret * 10) + digit;
	}
```